File: 40. Combination Sum II.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        candidates.sort()
        a = candidates
        self.dfs(a, 0, [], target, res)
        return res
        
    def dfs(self, a, idx, nums, target, res):
        if 0 == target:
            res.append(nums[:])
            return
        
        i = idx
        while i < len(a) and target >=  a[i]:
            nums.append(a[i])
            self.dfs(a, i + 1, nums, target - a[i], res)
            nums.pop()
            while i+1 < len(a) and a[i+1] == a[i]:
                i += 1
                
            i += 1
        

File: go/src/typical_problems/组合/40. Combination Sum II/40. Combination Sum II.go
package main

import (
	"fmt"
	"sort"
)

func combinationSum2(candidates []int, target int) [][]int {
	var ans = [][]int{}
	n := len(candidates)
	sort.Ints(candidates)

	var backtrace func(int, int, []int)
	backtrace = func(idx int, s int, arr []int) {
		if s <= 0|| idx == n {
			if s == 0 {
				ans = append(ans, append([]int{}, arr...))
			}
			return
		}

		for i := idx; i < n; i++ {
			if i > idx && candidates[i] == candidates[i - 1] {
				continue
			}
			arr = append(arr, candidates[i])
			backtrace(i + 1, s - candidates[i], arr)	// 只能选一次
			arr = arr[:len(arr) - 1]
		}
	}

	backtrace(0, target, []int{})
	return ans
}

func main() {
	fmt.Println(combinationSum2([]int{10,1,2,7,6,1,5}, 8))
}


